umbral/pre.py

```diff
@@ -216,50 +224,66 @@ def split_rekey(priv_a: Union[UmbralPrivateKey, CurveBN],
     """
     params = params if params is not None else default_params()
 
-    if isinstance(priv_a, UmbralPrivateKey):
-        priv_a = priv_a.bn_key
+    g = params.g
+
+    if isinstance(privkey_a_bn, UmbralPrivateKey):
+        pubkey_a_point = privkey_a_bn.get_pubkey().point_key
+        privkey_a_bn = privkey_a_bn.bn_key
+    else:
+        pubkey_a_point = privkey_a_bn * g 
 
     if isinstance(pubkey_b_point, UmbralPublicKey):
         pubkey_b_point = pubkey_b_point.point_key
 
-    g = params.g
-    pubkey_a_point = priv_a * g
-
-    x = CurveBN.gen_rand(params.curve)
-    xcomp = x * g
-    d = CurveBN.hash(xcomp, pubkey_b_point, pubkey_b_point * x, params=params)
+    # 'ni' stands for 'Non Interactive'.
+    # This point is used as an ephemeral public key in a DH key exchange,
+    # and the resulting shared secret 'd' allows to make Umbral non-interactive
+    priv_ni = CurveBN.gen_rand(params.curve)
+    ni = priv_ni * g
+    d = CurveBN.hash(ni, pubkey_b_point, pubkey_b_point * priv_ni, params=params)
 
-    coeffs = [priv_a * (~d)]
+    coeffs = [privkey_a_bn * (~d)]
     coeffs += [CurveBN.gen_rand(params.curve) for _ in range(threshold - 1)]
 
     u = params.u
 
-    g_ab = priv_a * pubkey_b_point
+    # 'xcoord' stands for 'X coordinate'.
+    # This point is used as an ephemeral public key in a DH key exchange,
+    # and the resulting shared secret 'dh_xcoord' contributes to prevent 
+    # reconstruction of the re-encryption key without Bob's intervention
+    priv_xcoord = CurveBN.gen_rand(params.curve)
+    xcoord = priv_xcoord * g
+
+    dh_xcoord = priv_xcoord * pubkey_b_point
 
     blake2b = hashes.Hash(hashes.BLAKE2b(64), backend=backend)
-    blake2b.update(pubkey_a_point.to_bytes())
+    blake2b.update(xcoord.to_bytes())
     blake2b.update(pubkey_b_point.to_bytes())
-    blake2b.update(g_ab.to_bytes())
+    blake2b.update(dh_xcoord.to_bytes())
     hashed_dh_tuple = blake2b.finalize()
 
+    bn_size = CurveBN.get_size(params.curve)
+
     kfrags = []
     for _ in range(N):
-        id_kfrag = CurveBN.gen_rand(params.curve)
+        id = os.urandom(bn_size)
```

Comment:
Yay! +1 for nondeterminism!

Reply:
I guess I'm missing something here - how is this less deterministic than `BN_rand_range`?

Reply:
Oh, I'm just putting a general, positive comment on a PR. Nothing special. This makes the ID a bitstring and I think it's great we can use `os.urandom` here.